### PR TITLE
Determinant-based CI solver

### PR DESCRIPTION
**Pull request by lcyyork**
## Description
Add the determinant-based CI solver using `CI_Reference` and `SparseCISolver`. Start retiring the old `FCI_MO` code.
In general, this general DETCI code is much slower than FCI and Psi4's DETCI module because it is not vectorized.

This PR also fix the edge case where no orbital optimization is asked for MCSCF.

Several test cases added/changed:
- `detci-1`: DOCI test on HF molecule
- `detci-2`: GASCI on H2O
- `detci-3`: transition dipoles and oscillator strengths on butadiene Ag -> Bu
- `detci-4`: CASCI that uses `SparseList` to build sigma vector
- `casscf-opt-3`: frozen-core FCI optimization using `mcscf_two_step` code.
- `dsrg-mrpt2-8-sa`: add tests for dipole moments
- `sa-casscf-2`: loose convergence threshold from 1.0e-15 -> 1.0e-12
- `sa-gasscf-2`: some clean up
- `sa-gasscf-dsrg-mrpt2-1`: use `detci` instead of `aci` for gasscf

## User Notes
- [x] All currently supported `FCI_MO` functionalities are also available in `DETCI`.
- [x] Add general functions to dump/read CI wave function to disk for `ActiveSpaceSolver` and `ActiveSpaceMethod` to be inherited by the children classes.
- [x] Add dump and read (as initial guess) CI wave functions to and from disk for `DETCI`.
- [x] Fix a bug of CI_RDMs when computing transition RDMs.
- [x] Implement `ActiveSpaceSolver::compute_fosc_same_orbs`.
- [x] Implement `ActiveSpaceMethod::compute_oscillator_strength_same_orbs` and `ActiveSpaceMethod::compute_transition_dipole_same_orbs`.
- [x] Change the key pushed to Psi4 environment globals. For example, if `"ENERGY ROOT 0 1A1"` has been pushed to Psi4 globals, we will call the current energy `"ENERGY ROOT 0 1A1 ENTRY 1"` and the old becomes `"ENERGY ROOT 0 1A1 ENTRY 0"`. See test cases `detci-2` and `dsrg-mrpt2-8-sa`.
- [x] Change the format of `guess_` in `SparseCISolver ` to `std::vector<std::vector<std::pair<size_t, double>>>`:  nroot of guess size of (id, coefficent)
- [x] Overload `operator[]` for `DeterminantHashVec`, which now also supports range-based loops by adding `begin()` and `end()`.
- [x] Add another two ways of string representation of `StateInfo`. The minimal one contains only the state irrep and multiplicity information. The other is used for file names (e.g., low spin triplet A1 state: `m1.z0.h0.g_gasmin.g_gasmax`).

## Checklist
- [x] Added tests of new features
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [x] Checked for consistency in the formatting of the output file
- [x] Documented new features in the manual
- [x] Ready to review
- [ ] Ready to go!

### CODE REVIEW

File: src/base_classes/active_space_solver.cc [hunk 1]
```diff
@@ -591,6 +602,26 @@ RDMs ActiveSpaceSolver::compute_avg_rdms_ms_avg(
     return RDMs(true, g1a, g2ab, g3aab);
 }
 
+std::map<StateInfo, std::string> ActiveSpaceSolver::dump_wave_function() {
+    std::map<StateInfo, std::string> out;
+
+    std::string method_lowercase(method_);
+    std::transform(method_lowercase.begin(), method_lowercase.end(), method_lowercase.begin(),
```
Comment (fevangelista) [inline src/base_classes/active_space_solver.cc:609]: There is a function to do this in `helpers/string_algorithms.h`

File: src/base_classes/active_space_solver.cc [hunk 2]
```diff
@@ -591,6 +602,26 @@ RDMs ActiveSpaceSolver::compute_avg_rdms_ms_avg(
     return RDMs(true, g1a, g2ab, g3aab);
 }
 
+std::map<StateInfo, std::string> ActiveSpaceSolver::dump_wave_function() {
+    std::map<StateInfo, std::string> out;
+
+    std::string method_lowercase(method_);
```
Comment (fevangelista) [inline src/base_classes/active_space_solver.cc:608]:
```suggestion
    std::string method_lowercase = lower_string(method_);
```

File: src/base_classes/active_space_solver.h [hunk 3]
```diff
@@ -111,6 +112,23 @@ class ActiveSpaceSolver {
         as_ints_ = as_ints;
     }
 
+    /// Save the wave function
```
Comment (fevangelista) [inline src/base_classes/active_space_solver.h:115]: Can you say what does this function return?

File: src/base_classes/active_space_solver.h [hunk 4]
```diff
@@ -111,6 +112,23 @@ class ActiveSpaceSolver {
...
+    std::map<StateInfo, std::string> dump_wave_function();
+
+    /// Read the wave function from file as initial guess
+    void read_wave_function(const std::map<StateInfo, std::string>& state_filename_map) {
+        state_filename_map_ = state_filename_map;
+    };
+
+    /// Set energy convergence
```
Comment (fevangelista) [inline src/base_classes/active_space_solver.h:123]: I am surprised you needed to add these. Is there another mechanism to set these variables?

File: src/casscf/mcscf_2step.h [hunk 5]
```diff
@@ -129,6 +129,11 @@ class MCSCF_2STEP {
     /// Orbital gradient convergence criteria
     double g_conv_;
 
+    /// Energy convergence for Davidson-Liu
+    double dl_e_conv_ = 1.0e-4;
```
Comment (fevangelista) [inline src/casscf/mcscf_2step.h:133]: These are very loose defaults. Do you expect these would be useful or they always get override?

File: src/casscf/mcscf_2step.h [hunk 6]
```diff
@@ -138,7 +143,10 @@ class MCSCF_2STEP {
     /// Solve CI coefficients for the current orbitals
     std::unique_ptr<ActiveSpaceSolver>
     diagonalize_hamiltonian(std::shared_ptr<ActiveSpaceIntegrals> fci_ints, const int print,
-                            double& e_c);
+                            const bool do_dipole, double& e_c);
```
Comment (fevangelista) [inline src/casscf/mcscf_2step.h:146]: what is the last argument here? `double& e_c`? If this is an energy why not just grab it back from the returned `ActiveSpaceSolver` object? This mechanism of getting variables updated is not really ideal, unless there is a performance issue, which should not be the case here.

File: src/ci_rdm/ci_rdms.cc [hunk 7]
```diff
@@ -968,7 +978,7 @@ void CI_RDMS::compute_3rdm_op(std::vector<double>& tprdm_aaa, std::vector<double
                 const double& sign_stu = std::get<1>(detI) > 0.0 ? 1.0 : -1.0;
                 const size_t& I = std::get<0>(detI);
 
-                double el = evecs_->get(I, root1_) * evecs_->get(J, root2_) * sign_pqr * sign_stu;
+                double el = evecs_->get(J, root1_) * evecs_->get(I, root2_) * sign_pqr * sign_stu;
```
Comment (fevangelista) [inline src/ci_rdm/ci_rdms.cc:981]: Are there tests that check for these changes?

File: src/sci/detci.h [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    /// Transition RDMs override
+    std::vector<RDMs> transition_rdms(const std::vector<std::pair<size_t, size_t>>& root_list,
+                                      std::shared_ptr<ActiveSpaceMethod> method2,
+                                      int max_rdm_level) override;
+
+    /// Set options override
+    void set_options(std::shared_ptr<ForteOptions> options) override;
+
+    /// Quiet mode (no printing, for use with CASSCF)
```
Comment (fevangelista) [inline src/sci/detci.h:47]: Should these functions be added to `ActiveSpaceMethods`?

File: src/sci/detci.h [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    /// ForteOptions
+    std::shared_ptr<ForteOptions> options_;
+
+    /// Start up function
+    void startup();
+
+    /// Quiet printing
+    bool quiet_ = false;
+
+    /// Number of active orbitals
```
Comment (fevangelista) [inline src/sci/detci.h:79]: In the past when people added code in this class there was a lot of variable repetition, meaning there were variables with the same name or functionality in both the base and derived classes. Please make sure to eliminate and rename any variable here that is already implemented in the base class. Otherwise there is potential for major confusion and errors.

File: src/sci/detci.h [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    /// CI_Reference object
+    std::shared_ptr<CI_Reference> ci_ref_;
+    /// The determinant space
+    DeterminantHashVec p_space_;
+    /// Build determinant space
+    void build_determinant_space();
+
+    /// State label
+    std::string state_label_;
+    /// Multiplicity
```
Comment (fevangelista) [inline src/sci/detci.h:95]: Use the state information object instead of these.

File: src/sparse_ci/ci_reference.cc [hunk 11]
```diff
@@ -518,6 +504,24 @@ void CI_Reference::build_cas_reference(std::vector<Determinant>& ref_space) {
     outfile->Printf("\n  Reference generated from %d MOs", na);
 }
 
+void CI_Reference::build_cas_reference_full(std::vector<Determinant>& ref_space) {
```
Comment (fevangelista) [inline src/sparse_ci/ci_reference.cc:507]:
```suggestion
void CI_Reference::build_cas_reference(std::vector<Determinant>& ref_space) {
```
I am not sure we need the work `full`, unless it distinguishes this function from another one.

**Review by fevangelista — CHANGES_REQUESTED** [review]
Thanks for this PR and for making things a bit more uniform. I made several suggestions but we might need some more work in the future to clean this up.

File: src/base_classes/active_space_solver.h (hunk 4 above)
Comment (lcyyork) [reply src/base_classes/active_space_solver.h:123]: I am not sure what is the best way, but the `ActiveSpaceMethod` has them fixed in the [header](https://github.com/evangelistalab/forte/pull/227/files/11248643ecc15886d5e279b140bd25d65e44708c#diff-aa6bb529a4e6b980f3e07a088a6f95163d4ffcfb44da9f5cb8f6d00cca4da5deR208) and requires the children to override the `set_options` function. I add this simply because `ActiveSpaceSolver` is the user interface and during the MCSCF optimization, we probably don't need to converge things too tightly.

File: src/casscf/mcscf_2step.h (hunk 5 above)
Comment (lcyyork) [reply src/casscf/mcscf_2step.h:133]: These DL convergence thresholds will be tightened along the way through orbital optimization.

File: src/casscf/mcscf_2step.h (hunk 6 above)
Comment (lcyyork) [reply src/casscf/mcscf_2step.h:146]: Yes, `e_c` is the averaged energy. I guess I did not want to repeat the line of computing averaged energy, but I have made the changes.

File: src/ci_rdm/ci_rdms.cc (hunk 7 above)
Comment (lcyyork) [reply src/ci_rdm/ci_rdms.cc:981]: No. The old implementation assumes Hermitian RDMs, which (I think) are not guaranteed for transition RDMs. I will see if I can come up with a test case for transition RDMs

File: src/sci/detci.h (hunk 8 above)
Comment (lcyyork) [reply src/sci/detci.h:47]: Yes, moved to ActiveSpaceMethod

File: src/sci/detci.h (hunk 10 above)
Comment (lcyyork) [reply src/sci/detci.h:95]: These are actually read from `state_` ([here](https://github.com/evangelistalab/forte/pull/227/files/11248643ecc15886d5e279b140bd25d65e44708c#diff-8c55bdbc7eee5cf3882609c3a7791672ea21bf8744d6a96d0d141fb8c575edb0R38)) from `ActiveSpaceMethod`.

File: src/sci/detci.h (hunk 9 above)
Comment (lcyyork) [reply src/sci/detci.h:79]: They are set from `mo_space_info_` in `ActiveSpaceMethods`.

File: src/sparse_ci/ci_reference.cc (hunk 11 above)
Comment (lcyyork) [reply src/sparse_ci/ci_reference.cc:507]: Yes, there is a distinction between the two. The original one only selects up to the `subspace_size_`. The full one gives all the CAS determinants.

File: src/base_classes/active_space_solver.cc (hunk 1 above)
Comment (lcyyork) [reply src/base_classes/active_space_solver.cc:609]: Thanks, changes made.

File: src/base_classes/active_space_solver.h (hunk 3 above)
Comment (lcyyork) [reply src/base_classes/active_space_solver.h:115]: It returns a map from state to the filename of the wave function

File: src/base_classes/active_space_solver.cc (hunk 2 above)
Comment (lcyyork) [reply src/base_classes/active_space_solver.cc]: Changes made

File: src/ci_rdm/ci_rdms.cc (hunk 7 above)
Comment (lcyyork) [reply src/ci_rdm/ci_rdms.cc:981]: So I added the tests of DETCI dipoles in `dsrg-mrpt2-8-sa`, which should tell the difference.

File: src/casscf/mcscf_2step.h (hunk 5 above)
Comment (fevangelista) [reply src/casscf/mcscf_2step.h:133]: Ok, it might be a good idea to say "These DL convergence thresholds will be tighten up along the way through orbital optimization" in the description of this variable.

Comment (lcyyork) [reply src/casscf/mcscf_2step.h]: I changed these convergence thresholds to temporary variables before iterations.

**Comment by lcyyork** [comment]
The last commit simply lowers the convergence thresholds because of the prior azure tests failed to converge SCF energy to 1.0e-15.

File: register_forte_options.py [hunk 12]
```diff
@@ -919,6 +926,8 @@ def register_casscf_options(options):
     options.add_bool("CASSCF_DEBUG_PRINTING", False,
                      "Enable debug printing if True")
 
+    options.add_bool("CASSCF_DONT_ORBITAL", False, "No orbital optimization if true")
```
Comment (fevangelista) [inline register_forte_options.py:929]: maybe use something that better explains what this does as:
```suggestion
    options.add_bool("CASSCF_NO_ORBOPT", False, "No orbital optimization if true")
```

File: register_forte_options.py [hunk 13]
```diff
@@ -539,16 +549,6 @@ def register_aci_options(options):
 
     options.add_bool("PI_ACTIVE_SPACE", False, "Active space type?")
 
-    options.add_bool("SPIN_MAT_TO_FILE", False,
```
Comment (fevangelista) [inline register_forte_options.py:542]: Are these unused or replaced by other options?

**Review by fevangelista — APPROVED** [review]
I left a question and a suggestion. I can merge when you give me the OK.

File: register_forte_options.py (hunk 13 above)
Comment (lcyyork) [reply register_forte_options.py:542]: These options are repeated just around line 535.

File: register_forte_options.py (hunk 12 above)
Comment (lcyyork) [reply register_forte_options.py]: Changes made.

**Comment by lcyyork** [comment]
@fevangelista I think there might be a problem of directly read previous wave function as initial guess. Let me investigate the issue a little bit.

**Comment by fevangelista** [comment]
Ooops. I missed your last comment. Can you put the fix in a new PR?

**Comment by lcyyork** [comment]
@fevangelista OK, I do find the DL solver is somewhat sensitive for the guess size when computing excited states. May be it is relevant to the issue of manually setting the initial guess vectors.

```
import forte
memory 1 gb

molecule octatetraene{
0 1
H  0.971328  1.220141  0.000000
H -0.971328 -1.220141  0.000000
H -2.098090  0.984719  0.000000
H  2.098090 -0.984719  0.000000
H -0.146884  3.418505  0.000000
H  0.146884 -3.418505  0.000000
H -2.193473  4.766086  0.000000
H  2.193473 -4.766086  0.000000
H -3.225698  3.230501  0.000000
H  3.225698 -3.230501  0.000000
C  0.000000  0.721498  0.000000
C  0.000000 -0.721498  0.000000
C  1.125020 -1.479523  0.000000
C -1.125020  1.479523  0.000000
C  1.121077 -2.928812  0.000000
C -1.121077  2.928812  0.000000
C  2.237388 -3.682282  0.000000
C -2.237388  3.682282  0.000000
}

set {
  reference           rhf
  scf_type            df
  basis               def2-svp
  maxiter             300
  e_convergence       12
  d_convergence       10
  docc                [13,2,2,12]
}
Escf, wfn = energy('scf', return_wfn=True)

set forte {
  active_space_solver  fci
  int_type             df
  restricted_docc      [13,0,0,12]
  active               [0,4,4,0]
  frozen_uocc          [0,0,0,0]
  e_convergence        8
  r_convergence        6
  fci_maxiter          1000
  avg_state            [[0,1,5], [3,1,3]]
  NTRIAL_PER_ROOT 3  # change this
}
Ecas, wfn = energy('forte', ref_wfn=wfn, return_wfn=True)
```

5 and more give the correct energies.

1:
```
    Multi.(2ms)  Irrep.  No.               Energy
    ---------------------------------------------
       1  (  0)    Ag     0     -308.554608905439
       1  (  0)    Ag     1     -308.371770499882
       1  (  0)    Ag     2     -308.304411601959
       1  (  0)    Ag     3     -308.284072668911
       1  (  0)    Ag     4     -308.265980248601
    ---------------------------------------------
       1  (  0)    Bu     0     -308.329300595331
       1  (  0)    Bu     1     -308.328577485564
       1  (  0)    Bu     2     -308.323643769769
    ---------------------------------------------
```

2:
```
    Multi.(2ms)  Irrep.  No.               Energy
    ---------------------------------------------
       1  (  0)    Ag     0     -308.554608905440
       1  (  0)    Ag     1     -308.371770499882
       1  (  0)    Ag     2     -308.304411601959
       1  (  0)    Ag     3     -308.265980248600
       1  (  0)    Ag     4     -308.240241310634
    ---------------------------------------------
       1  (  0)    Bu     0     -308.329300595331
       1  (  0)    Bu     1     -308.323643769769
       1  (  0)    Bu     2     -308.245971430747
    ---------------------------------------------
```

3:
```
    Multi.(2ms)  Irrep.  No.               Energy
    ---------------------------------------------
       1  (  0)    Ag     0     -308.554608905440
       1  (  0)    Ag     1     -308.371770499882
       1  (  0)    Ag     2     -308.304411601959
       1  (  0)    Ag     3     -308.265980248601
       1  (  0)    Ag     4     -308.240241310634
    ---------------------------------------------
       1  (  0)    Bu     0     -308.329300595330
       1  (  0)    Bu     1     -308.323643769769
       1  (  0)    Bu     2     -308.245971430746
    ---------------------------------------------
```

5:
```
    Multi.(2ms)  Irrep.  No.               Energy
    ---------------------------------------------
       1  (  0)    Ag     0     -308.554608905440
       1  (  0)    Ag     1     -308.371770499883
       1  (  0)    Ag     2     -308.304411601960
       1  (  0)    Ag     3     -308.265980248601
       1  (  0)    Ag     4     -308.240241310634
    ---------------------------------------------
       1  (  0)    Bu     0     -308.329300595331
       1  (  0)    Bu     1     -308.323643769768
       1  (  0)    Bu     2     -308.239781796223
    ---------------------------------------------
```

10:
```
    Multi.(2ms)  Irrep.  No.               Energy
    ---------------------------------------------
       1  (  0)    Ag     0     -308.554608905440
       1  (  0)    Ag     1     -308.371770499882
       1  (  0)    Ag     2     -308.304411601959
       1  (  0)    Ag     3     -308.265980248601
       1  (  0)    Ag     4     -308.240241310633
    ---------------------------------------------
       1  (  0)    Bu     0     -308.329300595331
       1  (  0)    Bu     1     -308.323643769769
       1  (  0)    Bu     2     -308.239781796224
    ---------------------------------------------
```
